src/library/filescanner_ffmpeg.c

```diff
@@ -778,3 +802,234 @@ scan_metadata_ffmpeg(struct media_file_info *mfi, const char *file)
+      if (pkt.stream_index >= number_of_streams || stream_mapping[pkt.stream_index] < 0)
+	{
+	  av_packet_unref (&pkt);
+	  continue;
+	}
+
+      pkt.stream_index = stream_mapping[pkt.stream_index];
+      out_stream = out_fmt_ctx->streams[pkt.stream_index];
+
+      /* copy packet */
+      pkt.pts = av_rescale_q_rnd (pkt.pts, in_stream->time_base, out_stream->time_base, AV_ROUND_NEAR_INF | AV_ROUND_PASS_MINMAX);
+      pkt.dts = av_rescale_q_rnd (pkt.dts, in_stream->time_base, out_stream->time_base, AV_ROUND_NEAR_INF | AV_ROUND_PASS_MINMAX);
+      pkt.duration = av_rescale_q (pkt.duration, in_stream->time_base, out_stream->time_base);
+      pkt.pos = -1;
+
+      ret = av_interleaved_write_frame (out_fmt_ctx, &pkt);
+      if (ret < 0)
+	{
+	  DPRINTF(E_LOG, L_SCAN, "Error muxing pkt for rating '%s' - %s\n", in_fmt_ctx->url, av_err2str(ret));
+	  break;
+	}
+      av_packet_unref (&pkt);
+    }
+  av_write_trailer (out_fmt_ctx);
+
+end:
+  if (out_fmt_ctx && !(out_fmt_ctx->oformat->flags & AVFMT_NOFILE))
+    {
+      avio_closep (&out_fmt_ctx->pb);
+    }
+  avformat_free_context (out_fmt_ctx);
+  av_freep (&stream_mapping);
+  if (ret < 0 && ret != AVERROR_EOF)
+    {
+      return -1;
+    }
+  return 0;
+}
+
+int
+filescanner_ffmpeg_write_rating(const struct media_file_info *mfi)
+{
+  int ret;
+  char rating[5] = { '\0' };
+  char dest[PATH_MAX];
+  int i;
+  bool supported = false;
+  AVDictionaryEntry  *entry;
+
+  AVFormatContext *ctx = NULL;
+  if ( (ret = avformat_open_input(&ctx, mfi->path, NULL, NULL)) != 0)
+    {
+      DPRINTF(E_LOG, L_SCAN, "failed to open library file for rating metadata update '%s' - %s\n", mfi->path, av_err2str(ret));
+      return ENOENT;
+    }
+
+  for (i=0; i<ctx->nb_streams && !supported; ++i)
+  {
+      if (ctx->streams[i]->codecpar->codec_type != AVMEDIA_TYPE_AUDIO)
+        continue;
+
+      switch (ctx->streams[i]->codecpar->codec_id)
+      {
+	// ffmpeg's metadata update are limited - some formats do not support
+	// rating update even though the write completes; keep this in sync with
+	// supported formats
+        case AV_CODEC_ID_FLAC:
+        case AV_CODEC_ID_MP3:
+	  supported = true;
+	  break;
+
+        default:
+	  DPRINTF(E_WARN, L_SCAN, "unsupported metadata update for 'rating' on '%s' (%s) - skipping\n", mfi->path, avcodec_get_name(ctx->streams[i]->codecpar->codec_id));
+      }
+  }
+
+  if (!supported)
+    goto end;
+
+  safe_snprintf_cat(rating, 4, "%d", mfi->rating);
+
+  // Save a potential write if metadata on the underlying file matches requested rating
+  entry = av_dict_get(ctx->metadata, "rating", NULL, 0);
+  if (entry == NULL || (entry && entry->value == NULL) || (entry && strcmp(entry->value, rating) != 0) )
+    {
+      av_dict_set(&ctx->metadata, "rating", rating, 0);
+      DPRINTF(E_LOG, L_SCAN, "updating rating to %s on '%s'\n", rating, mfi->path);
+
+      sprintf(dest, "%s-%d.metadata", ctx->url, getpid());
```

Comment:
we do a `rename` later to replace the original file, which means it needs to be on the same device.  the `mkstemps()` template creates a path in the same place as the original file to avoid this limitation from `rename()`